### ISSUE TITLE
TestWatchEtcdError fairly consistently timing out

### ISSUE THREAD

**Issue by ixdy** [kubernetes/kubernetes#18914]
Not 100%, but it's happening frequently enough to block merges

```
=== RUN   TestWatchEtcdError
panic: test timed out after 5m0s

goroutine 950 [running]:
testing.startAlarm.func1()
	/tmp/workdir/go/src/testing/testing.go:703 +0x16b
created by time.goFunc
	/tmp/workdir/go/src/time/sleep.go:129 +0x6e

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x1b1d5f0, 0x22074c0, 0x1e, 0x1e, 0x1)
	/tmp/workdir/go/src/testing/testing.go:562 +0xafa
testing.(*M).Run(0xc82005ff10, 0x0)
	/tmp/workdir/go/src/testing/testing.go:494 +0xe5
main.main()
	k8s.io/kubernetes/pkg/storage/etcd/_test/_testmain.go:112 +0x210

goroutine 17 [syscall, 4 minutes, locked to thread]:
runtime.goexit()
	/tmp/workdir/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 6 [chan receive]:
github.com/golang/glog.(*loggingT).flushDaemon(0x22292e0)
	/workspace/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:879 +0x8c
created by github.com/golang/glog.init.1
	/workspace/kubernetes/Godeps/_workspace/src/github.com/golang/glog/glog.go:410 +0x2ba

goroutine 939 [select, 4 minutes]:
github.com/coreos/etcd/etcdserver/etcdhttp.handleKeyWatch(0x7fa43e617848, 0xc8206562a0, 0x7fa43e5d7b08, 0xc8204cc0b0, 0x7fa43e61c090, 0xc820bb4180, 0xc820bd6000, 0x7fa43e5d7500, 0xc820c342c0)
	/workspace/kubernetes/Godeps/_workspace/src/github.com/coreos/etcd/etcdserver/etcdhttp/client.go:626 +0xadb
github.com/coreos/etcd/etcdserver/etcdhttp.(*keysHandler).ServeHTTP(0xc82039d130, 0x7fa43e5d7b08, 0xc8204cc0b0, 0xc820bfe000)
	/workspace/kubernetes/Godeps/_workspace/src/github.com/coreos/etcd/etcdserver/etcdhttp/client.go:158 +0xbe2
net/http.(*ServeMux).ServeHTTP(0xc82058d4a0, 0x7fa43e5d7b08, 0xc8204cc0b0, 0xc820bfe000)
	/tmp/workdir/go/src/net/http/server.go:1699 +0x213
github.com/coreos/etcd/etcdserver/etcdhttp.requestLogger.func1(0x7fa43e5d7b08, 0xc8204cc0b0, 0xc820bfe000)
	/workspace/kubernetes/Godeps/_workspace/src/github.com/coreos/etcd/etcdserver/etcdhttp/http.go:85 +0x2a9
net/http.HandlerFunc.ServeHTTP(0xc82014c520, 0x7fa43e5d7b08, 0xc8204cc0b0, 0xc820bfe000)
	/tmp/workdir/go/src/net/http/server.go:1422 +0x48
net/http/httptest.(*waitGroupHandler).ServeHTTP(0xc82014c560, 0x7fa43e5d7b08, 0xc8204cc0b0, 0xc820bfe000)
	/tmp/workdir/go/src/net/http/httptest/server.go:200 +0xff
net/http.serverHandler.ServeHTTP(0xc820674060, 0x7fa43e5d7b08, 0xc8204cc0b0, 0xc820bfe000)
	/tmp/workdir/go/src/net/http/server.go:1862 +0x207
net/http.(*conn).serve(0xc820654000)
	/tmp/workdir/go/src/net/http/server.go:1361 +0x117d
created by net/http.(*Server).Serve
	/tmp/workdir/go/src/net/http/server.go:1910 +0x465

goroutine 984 [select, 4 minutes]:
k8s.io/kubernetes/pkg/storage/etcd.(*etcdWatcher).translate(0xc820aaa000)
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_watcher.go:221 +0x8ec
created by k8s.io/kubernetes/pkg/storage/etcd.newEtcdWatcher
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_watcher.go:137 +0x46c

goroutine 985 [select, 4 minutes]:
github.com/coreos/etcd/client.(*simpleHTTPClient).Do(0xc820aa2120, 0x7fa43e5d79d0, 0xc82016e580, 0x7fa43fe67a60, 0xc82016e390, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/workspace/kubernetes/Godeps/_workspace/src/github.com/coreos/etcd/client/client.go:439 +0x7b5
github.com/coreos/etcd/client.(*redirectFollowingHTTPClient).Do(0xc82014c940, 0x7fa43e5d79d0, 0xc82016e580, 0x7fa43fe67a60, 0xc82016e390, 0xc82014c587, 0x0, 0x0, 0x0, 0x0, ...)
	/workspace/kubernetes/Godeps/_workspace/src/github.com/coreos/etcd/client/client.go:474 +0x19f
github.com/coreos/etcd/client.(*httpClusterClient).Do(0xc82039c0f0, 0x7fa43e5d79d0, 0xc82016e580, 0x7fa43fe67a60, 0xc82016e390, 0xc8200126a0, 0x0, 0x0, 0x0, 0x0, ...)
	/workspace/kubernetes/Godeps/_workspace/src/github.com/coreos/etcd/client/client.go:275 +0x6d4
github.com/coreos/etcd/client.(*httpWatcher).Next(0xc82016e380, 0x7fa43e5d79d0, 0xc82016e580, 0xc82016e580, 0x0, 0x0)
	/workspace/kubernetes/Godeps/_workspace/src/github.com/coreos/etcd/client/keys.go:435 +0xf6
k8s.io/kubernetes/pkg/storage/etcd.(*etcdWatcher).etcdWatch(0xc820aaa000, 0x7fa43e6177d8, 0xc8200126a0, 0x7fa43e5d76a8, 0xc82014c5e0, 0xc82014c880, 0x12, 0x4)
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_watcher.go:164 +0x424
created by k8s.io/kubernetes/pkg/storage/etcd.(*etcdHelper).Watch
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_helper.go:271 +0x2cd

goroutine 1048 [chan receive, 4 minutes]:
k8s.io/kubernetes/pkg/storage/etcd.TestWatchEtcdError(0xc8208d0cf0)
	/workspace/kubernetes/_output/local/go/src/k8s.io/kubernetes/pkg/storage/etcd/etcd_watcher_test.go:230 +0x485
testing.tRunner(0xc8208d0cf0, 0x22076d0)
	/tmp/workdir/go/src/testing/testing.go:456 +0xdd
created by testing.RunTests
	/tmp/workdir/go/src/testing/testing.go:561 +0xaa4

goroutine 975 [select, 4 minutes]:
net/http.(*persistConn).writeLoop(0xc820aaa160)
	/tmp/workdir/go/src/net/http/transport.go:1009 +0x4ee
created by net/http.(*Transport).dialConn
	/tmp/workdir/go/src/net/http/transport.go:686 +0x11e5

goroutine 974 [select, 4 minutes]:
net/http.(*persistConn).readLoop(0xc820aaa160)
	/tmp/workdir/go/src/net/http/transport.go:976 +0xe8f
created by net/http.(*Transport).dialConn
	/tmp/workdir/go/src/net/http/transport.go:685 +0x11c0

goroutine 1059 [IO wait, 4 minutes]:
net.runtime_pollWait(0x7fa43c239e40, 0x72, 0x0)
	/tmp/workdir/go/src/runtime/netpoll.go:157 +0x63
net.(*pollDesc).Wait(0xc82029a220, 0x72, 0x0, 0x0)
	/tmp/workdir/go/src/net/fd_poll_runtime.go:73 +0x56
net.(*pollDesc).WaitRead(0xc82029a220, 0x0, 0x0)
	/tmp/workdir/go/src/net/fd_poll_runtime.go:78 +0x44
net.(*netFD).Read(0xc82029a1c0, 0xc82053f000, 0x1000, 0x1000, 0x0, 0x7fa43fe5a050, 0xc820012150)
	/tmp/workdir/go/src/net/fd_unix.go:232 +0x27b
net.(*conn).Read(0xc820ab6000, 0xc82053f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/net/net.go:172 +0x121
net.(*TCPConn).Read(0xc820ab6000, 0xc82053f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:74 +0x7d
net/http.noteEOFReader.Read(0x7fa43fe66af8, 0xc820ab6000, 0xc820aaa1b8, 0xc82053f000, 0x1000, 0x1000, 0x44f350, 0x0, 0x0)
	/tmp/workdir/go/src/net/http/transport.go:1370 +0x75
net/http.(*noteEOFReader).Read(0xc8204e6060, 0xc82053f000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
	<autogenerated>:126 +0xf8
bufio.(*Reader).fill(0xc8208c6060)
	/tmp/workdir/go/src/bufio/bufio.go:97 +0x365
bufio.(*Reader).ReadSlice(0xc8208c6060, 0xa, 0x0, 0x0, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/bufio/bufio.go:328 +0x5a9
net/http/internal.readLine(0xc8208c6060, 0x0, 0x0, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/net/http/internal/chunked.go:110 +0x62
net/http/internal.(*chunkedReader).beginChunk(0xc82029e930)
	/tmp/workdir/go/src/net/http/internal/chunked.go:47 +0x5b
net/http/internal.(*chunkedReader).Read(0xc82029e930, 0xc8200c2000, 0x200, 0x200, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/net/http/internal/chunked.go:77 +0x11b
net/http.(*body).readLocked(0xc820bb4240, 0xc8200c2000, 0x200, 0x200, 0xc820bb4290, 0x0, 0x0)
	/tmp/workdir/go/src/net/http/transfer.go:632 +0x114
net/http.(*body).Read(0xc820bb4240, 0xc8200c2000, 0x200, 0x200, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/net/http/transfer.go:624 +0x158
net/http.(*bodyEOFSignal).Read(0xc820bb4280, 0xc8200c2000, 0x200, 0x200, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/net/http/transport.go:1296 +0x2e0
bytes.(*Buffer).ReadFrom(0xc820955e40, 0x7fa43c519058, 0xc820bb4280, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/bytes/buffer.go:173 +0x415
io/ioutil.readAll(0x7fa43c519058, 0xc820bb4280, 0x200, 0x0, 0x0, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/io/ioutil/ioutil.go:33 +0x1a1
io/ioutil.ReadAll(0x7fa43c519058, 0xc820bb4280, 0x0, 0x0, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/io/ioutil/ioutil.go:42 +0x72
github.com/coreos/etcd/client.(*simpleHTTPClient).Do.func3(0xc820194720, 0xc8200d4230, 0xc8208d0000, 0xc8200104e0)
	/workspace/kubernetes/Godeps/_workspace/src/github.com/coreos/etcd/client/client.go:435 +0x88
created by github.com/coreos/etcd/client.(*simpleHTTPClient).Do
	/workspace/kubernetes/Godeps/_workspace/src/github.com/coreos/etcd/client/client.go:437 +0x55b

goroutine 861 [IO wait, 4 minutes]:
net.runtime_pollWait(0x7fa43fe652e8, 0x72, 0x0)
	/tmp/workdir/go/src/runtime/netpoll.go:157 +0x63
net.(*pollDesc).Wait(0xc820174060, 0x72, 0x0, 0x0)
	/tmp/workdir/go/src/net/fd_poll_runtime.go:73 +0x56
net.(*pollDesc).WaitRead(0xc820174060, 0x0, 0x0)
	/tmp/workdir/go/src/net/fd_poll_runtime.go:78 +0x44
net.(*netFD).Read(0xc820174000, 0xc820b3a000, 0x8000, 0x8000, 0x0, 0x7fa43fe5a050, 0xc820012150)
	/tmp/workdir/go/src/net/fd_unix.go:232 +0x27b
net.(*conn).Read(0xc8206e2000, 0xc820b3a000, 0x8000, 0x8000, 0x16e9900, 0x0, 0x0)
	/tmp/workdir/go/src/net/net.go:172 +0x121
net.(*TCPConn).Read(0xc8206e2000, 0xc820b3a000, 0x8000, 0x8000, 0x8000, 0x0, 0x0)
	<autogenerated>:74 +0x7d
io.copyBuffer(0x7fa43c375698, 0xc820ae2010, 0x7fa43fe66af8, 0xc8206e2000, 0xc820b3a000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/io/io.go:381 +0x282
io.Copy(0x7fa43c375698, 0xc820ae2010, 0x7fa43fe66af8, 0xc8206e2000, 0x0, 0x0, 0x0)
	/tmp/workdir/go/src/io/io.go:351 +0x79
net/http.(*conn).closeNotify.func1(0xc820ae2010, 0x7fa43fe66af8, 0xc8206e2000, 0xc820654000)
	/tmp/workdir/go/src/net/http/server.go:182 +0x72
created by net/http.(*conn).closeNotify
	/tmp/workdir/go/src/net/http/server.go:188 +0x4a3
FAIL	k8s.io/kubernetes/pkg/storage/etcd	300.095s
```

Unclear if related to #18913. #18635 makes me highly suspicious - looks like the failures started right around when it was merged.

@kubernetes/goog-testing @k8s-oncall @wojtek-t @timothysc @mikedanese

**Comment by timothysc** [comment]
@ixdy What time did the UT's start failing?   I'm guessing it was after the 1.5 update.

**Comment by timothysc** [comment]
@ixdy that test was originally disabled b/c of flakes that we seen and re-enabled on client update b/c we had not seen the flakes.  I submitted a PR to re-disable..

**Comment by ixdy** [comment]
The first instance of this test timing out occurred in kubernetes-test-go build 4751, which started at 5:59am PST this morning. This was one build after #18635 was merged, but well before #13838 was merged.

**Comment by ixdy** [comment]
Lowering priority since test is disabled, but we should fix and reenable it eventually.

**Comment by timothysc** [comment]
It was originally created against a fake server, when we enabled against a real server, there were no guarantees given from the coreos folks on the client side, we should probably add as a requirement for etcd3.

**Comment by lavalamp** [comment]
@timothysc can you clarify? What isn't guaranteed that needs to be? Was this test testing a useful thing or should we just delete it?

**Comment by timothysc** [comment]
@lavalamp I vote to delete it.  There is no time guarantee on when a client connection will detect a asynch server-termination, or net-split, unless we start tuning the tcp keepalive parameters.  Which is why the test flakes.

**Comment by lavalamp** [comment]
What loses coverage if we delete this test?

**Comment by timothysc** [comment]
The test was originally written to "try-to" detect when a server had terminated, but it was only tested against a fake server, never a real one.  So I don't think you're "loosing" any coverage in this case.  

I like the "idea" behind the test, but the time-bounds for UT too constraining.  Perhaps a better idea would be to convert to a different testing category which does not have that time-boxed restriction, such as e2e.

**Comment by davidopp** [comment]
@timothysc Is it OK if I close this and open a new issue for you to write an equivalent test using e2e?

**Comment by lavalamp** [comment]
I think there's already an e2e test that kills etcd? We should leave this open until we get a PR deleting the test, I'll make one later unless @timothysc beats me to it.

**Comment by lavalamp** [comment]
I'll remove the test.
